### PR TITLE
Allow specifying the generator class.

### LINKED ISSUE #113: Allow registering new `visit_xxx` for `SourceGenerator`
I ran into an issue where I defined a custom `AST` node, and when I tried rendering the source with `astor.to_source(node)`, it crashed saying `No defined handler for node of type xxx`.

I understand that `astor` would not know how to render some unknown `AST` node, but it would be nice if we could register a visit function, such that when `astor` encounters a unknown `AST` node, instead of crashing it checks the registry for the visit function.

There are at least two reasons for having a custom `AST` class:
* A user defined a truly new `AST` class.
* The user created a derived class of some known `AST` class, e.g. a `Attribute`. Because `obj.__class__.__name__` is used, even though it inherits from some known `AST` class, `astor` does not know what to do with it.

I understand why the first use case would not be convincing. But number two should be common enough that allowing to register one `AST` class name as an alias for another, or alternatively allowing to register an explicit visit function would be useful (or both).

My use case is as follows: I do the tree walk and then replace some nodes with a proxy node. The proxy node refers to the original node. However, later at my convenience, I replace the node the proxy refers to with another node. This allows me to switch out nodes without having to walk the tree. Currently, I monkey patched `astor` to be able to handle this as follows:

```py
def visit_ASTNodeRef(self, node, *largs, **kwargs):
    return self.visit(node.ref_node, *largs, **kwargs)
SourceGenerator.visit_ASTNodeRef = visit_ASTNodeRef
```

But, it would be nice if there was an official way of doing this. I can work on a PR if that sounds acceptable.

### PR DESCRIPTION
**Pull request by matham**
Fixes #113 
Adds the ability to specify the class to be used to generate the source code.

### CODE REVIEW

File: astor/code_gen.py [hunk 1]
```diff
@@ -46,9 +47,15 @@ def to_source(node, indent_with=' ' * 4, add_line_information=False,
...
+    `source_generator_class` defaults to `None`, and specifies the
+    class that will be instantiated and used to generate the source code.
+    When `None`, it will use `SourceGenerator`.
+
     """
-    generator = SourceGenerator(indent_with, add_line_information,
-                                pretty_string)
+    if source_generator_class is None:
+        source_generator_class = SourceGenerator
+    generator = source_generator_class(
```
Comment (berkerpeksag) [inline astor/code_gen.py]: Could you also raise a `TypeError` if `source_generator_class` is not a subclass of `SourceGenerator`?

**Comment by berkerpeksag** [comment]
> Would it make sense if I added maxline to the function signature and forwarded it to split_line?

Could you open a new issue to discuss this one? There is an open PR to refactor `astor/source_repr.py` and its friends, so maybe we can come up with a better solution.

**Comment by matham** [comment]
I just wanted to update that I have not forgotten about the PR, it's just that I ended up using astor quite a bit in new code I wrote, so before I continued with the PR I wanted to get a clearer picture for myself how I was going to use astor to make sure my solution was good approach.

In the end, I just wrote my own version of `to_source` (https://github.com/matham/kivy/blob/ast_compile/kivy/lang/compiler/ast_parse.py#L320) because I needed further customization, so I'm wondering if I should continue with this PR, or if it's better that the user make their own `to_source`, like I did rather than making `to_source` more complex? Especially if the API will be changing as you mentioned?

If the underlying API used in `to_source` is available more clearly, maybe that's better than adding more stuff to `to_source`? But either way, I can add tests etc to this PR or drop it. Whichever you think is better.

**Comment by berkerpeksag** [comment]
I think this PR would still help users to customize source generator. I recently came across a blog post using astor (I can't read Chinese, but I can read Python :)): http://blog.soliloquize.org/2017/02/08/%E4%BD%BF%E7%94%A8ast%E7%A7%BB%E9%99%A4%E4%BB%A3%E7%A0%81%E4%B8%AD%E7%9A%84print%E8%AF%AD%E5%8F%A5/

With `source_generator_class`, the latest example can be written without needing to subclass `ast.NodeTransformer`:

```py
import astor

class RemovePrintGenerator(astor.code_gen.SourceGenerator):

    def visit_Call(self, node):
        if node.func.id != 'print':
            super().visit_Call(node)

print(astor.to_source(astor.parse_file('foo.py'), source_generator_class=RemovePrintGenerator))
```

**Comment by matham** [comment]
Thanks for finishing this off. I hadn't had the time to complete it.

**Pull request merged**